app/src/main/java/bg/zahov/app/ui/measures/info/MeasurementInfoViewModel.kt

```diff
@@ -1,100 +1,139 @@
-                    _state.postValue(
-                        State.Data(
-                            maxValue = measurementMax,
-                            minValue = measurementMin,
-                            suffix = measurementSuffix,
-                            entries = measureEntries
-                        )
+            // Collect selected measurement data and update the UI state with chart data.
+            measurementProvider.getSelectedMeasurement().collect {
+                var measureEntries: List<Entry> = listOf()
+                if (it.measurements.values.isNotEmpty()) {
+                    measureEntries = filterEntries(it.measurements.values.first())
+                }
+                _uiState.update { old ->
+                    old.copy(
+                        data = LineChartData(
+                            maxValue = measureEntries.maxOfOrNull { value -> value.y } ?: 0f,
+                            minValue = measureEntries.minOfOrNull { value -> value.y } ?: 0f,
+                            suffix = if (it.measurements.keys.isNotEmpty()) it.measurements.keys.first() else null,
+                            list = measureEntries
+                        ),
+                        loading = false
                     )
                 }
-            } catch (e: CriticalDataNullException) {
-                serviceError.initiateCountdown()
             }
         }
     }
 
+    /**
+     * Filters a list of measurements, grouping by the day of the month, and returning the most recent
+     * measurement per day.
+     *
+     * @param measurements The list of measurements to filter.
+     * @return A list of [Entry] objects representing the filtered data for the line chart.
+     */
     private fun filterEntries(measurements: List<Measurement>): List<Entry> {
-        val latestEntriesPerDay = mutableMapOf<Int, Measurement>()
-
-        for (measurement in measurements) {
-            if (latestEntriesPerDay[measurement.date.dayOfMonth] == null || latestEntriesPerDay[measurement.date.dayOfMonth]!!.date.isBefore(
-                    measurement.date
-                )
-            ) {
-                latestEntriesPerDay[measurement.date.dayOfMonth] = measurement
+        return measurements
+            .groupBy { it.date.dayOfMonth }
+            .mapNotNull { (_, dailyMeasurements) -> dailyMeasurements.maxByOrNull { it.date } }
+            .map { measurement ->
+                Entry().apply {
+                    x = measurement.date.dayOfMonth.toFloat()
+                    y = measurement.value.toFloat()
+                }
             }
+            .sortedBy { it.x }
+    }
+
+    /**
+     * Updates the history input state based on user input.
+     *
+     * @param text The new text input entered by the user.
+     */
+    fun onHistoryInputChange(text: String) {
+        if (text.isEmpty() || text.matches(Regex("^\\d+\$"))) {
+            _uiState.update { old -> old.copy(historyInput = text) }
```

Comment:
You should probably use a separate flow for the user input, otherwise your entire screen will recompose on every input change.

Reply:
made diffent flow